### DISCUSSION TITLE
Using table field as subfield

### DISCUSSION THREAD

**Discussion by lieto-dk** [Laravel-Backpack/community-forum#756] [Q&A (Help)] [answered]
Im using a table field inside a relationship as a subfield. When the table is stored in the json field in mysql the value is escaped. When i use the table field as a normal field the value is not escaped. It gives me some issues as Laravel cant cast the escaped string as array (as defined in my model). The biggest issue is that the same field in the database will work differently when i save it as as a subfield and when using it as a regular field. Anyone seen this issue and is there any options that could be used in the subfield  ?

**Comment by pxpm** [comment]
Hello @lieto-dk thanks for the question. 

I've never faced it, I don't recall using a table inside a repeatable. 

Can you provide more details so that I can try to reproduce the issue ? 

What kind of relationship is it ? What's the real problem ? Is that it stores it as a json array instead of a json string or the other way around ? 

Can you also tell me your backpack version with `php artisan backpack:version` ? 

Please provide as many details as you can so that I can easily reproduce your issue and try to help you. 

Cheers

**Reply by lieto-dk** [reply]
Hi, any update on this issue ? What should i do to get it adressed ? Let me know if you need additional information to look into it !

**Comment by lieto-dk** [comment]
Hi,
I have a category model with a has many relation to a categorylocale model
In the categorylocale model i have a table (locale_parameters) field that is a json field in the database (mysql 8) and which is casted as an array in the categorylocale model. (  'locale_parameters' => 'array',)
When i add the table field to the crud for the categorylocale model it saves the field without escaping characters ([{"key": "weekend_offer", "value": "1", "operator": "="}]), when use the same field in the subfields in the category crud (for the categorylocalemodel) i save the value is escaped ("[{\"key\":\"weekend_offer\",\"operator\":\"=\",\"value\":\"1\"}]").
Backpack handles as both values is shown correctly in the table in both the categorylocale field and the category subfield for categorylocale on the category crud.
If the value is escaped and i use categorylocale->locale_parameters in Laravel it comes out with an error saying that the content is not an array but a string (as Laravel cant parse the value of the fields as a array) ( foreach() argument must be of type array|object, string given ...).

Hope you can direct me in the right direction to solve this puzzle !

I use the following
### PHP VERSION:
PHP 8.1.13 (cli) (built: Nov 26 2022 14:07:36) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.13, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.13, Copyright (c), by Zend Technologies

### LARAVEL VERSION:
10.32.1.0

### BACKPACK PACKAGE VERSIONS:
backpack/basset: 1.2.1
backpack/crud: 6.3.0
backpack/generators: v4.0.2
backpack/pro: 2.0.18
backpack/theme-coreuiv2: 1.2.2
backpack/theme-tabler: 1.1.1

**Reply by lieto-dk** [reply]
Ups ... the editor here does not shown the escaped characters (press edit and you will see - its just a standard escaping with \ before " and so on.)

**Reply by lieto-dk** [reply]
Hi [pxpm], did you get a chance to look into this ? /jens

**Reply by lieto-dk** [reply]
HI pxpm, any update on this issue ?

**Reply by pxpm** [reply, marked as answer]
Hey @lieto-dk sorry it took me some time to get back here. 

As you already found out the problem is related with a double encoding. That happens because the table already sends the encoded information, and when your "model cast" for the repeatable container runs it double encodes the table field. 

A quick solution I can think of is to to "decode" the table fields in the repeatable container before the create/update processes, something like: 
```php
    // overwrite store and update methods
    use \Backpack\CRUD\app\Http\Controllers\Operations\CreateOperation {
        store as traitStore;
    }
    use \Backpack\CRUD\app\Http\Controllers\Operations\UpdateOperation {
        update as traitUpdate;
    }
    
    private function decodeTableFieldsFromRequest()
    {
        $request = $this->crud->getRequest();
        $repeatable = $request->get('repeatable');

        if(is_array($repeatable)) {
            array_map(function($item) {
                $item['feature'] = json_decode($item['feature'] ?? '', true);
                return $item;
            }, $repeatable);
        }
        $request->request->set('repeatable', $repeatable);
        $this->crud->setRequest($request);
    }

    public function update($id)
    {
        $this->decodeTableFieldsFromRequest();
        return $this->traitUpdate($id);
    }

    public function store()
    {
        $this->decodeTableFieldsFromRequest();
        return $this->traitStore();
   }
```

Added this to the docs too. https://backpackforlaravel.com/docs/6.x/crud-fields#using-the-table-in-a-repeatable-field
